File: src/tools/transitions.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { OBSWebSocketClient } from "../client.js";
import { z } from "zod";

export async function initialize(server: McpServer, client: OBSWebSocketClient): Promise<void> {
  // GetTransitionList tool
  server.tool(
    "obs-get-transition-list",
    "Get a list of available transitions in OBS",
    {},
    async () => {
      try {
        const transitions = await client.sendRequest("GetTransitionList");
        return {
          content: [
            {
              type: "text",
              text: JSON.stringify(transitions, null, 2)
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting transition list: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetCurrentTransition tool
  server.tool(
    "obs-get-current-transition",
    "Get the name of the currently active transition",
    {},
    async () => {
      try {
        const transition = await client.sendRequest("GetCurrentSceneTransition");
        return {
          content: [
            {
              type: "text",
              text: `Current transition: ${transition.transitionName} (${transition.transitionKind}) with duration: ${transition.transitionDuration}ms`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting current transition: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetCurrentTransition tool
  server.tool(
    "obs-set-current-transition",
    "Set the current transition in OBS",
    {
      transitionName: z.string().describe("The name of the transition to set as current")
    },
    async ({ transitionName }) => {
      try {
        await client.sendRequest("SetCurrentSceneTransition", { transitionName });
        return {
          content: [
            {
              type: "text",
              text: `Successfully set current transition to: ${transitionName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting current transition: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetTransitionDuration tool
  server.tool(
    "obs-get-transition-duration",
    "Get the duration of the current transition in milliseconds",
    {},
    async () => {
      try {
        const duration = await client.sendRequest("GetCurrentSceneTransitionDuration");
        return {
          content: [
            {
              type: "text",
              text: `Current transition duration: ${duration.transitionDuration}ms`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting transition duration: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetTransitionDuration tool
  server.tool(
    "obs-set-transition-duration",
    "Set the duration of the current transition in milliseconds",
    {
      duration: z.number().min(0).describe("The duration to set in milliseconds")
    },
    async ({ duration }) => {
      try {
        await client.sendRequest("SetCurrentSceneTransitionDuration", { transitionDuration: duration });
        return {
          content: [
            {
              type: "text",
              text: `Successfully set transition duration to: ${duration}ms`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting transition duration: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetTransitionKind tool
  server.tool(
    "obs-get-transition-kind",
    "Get the kind/type of the current transition",
    {},
    async () => {
      try {
        const transition = await client.sendRequest("GetCurrentSceneTransition");
        return {
          content: [
            {
              type: "text",
              text: `Current transition kind: ${transition.transitionKind}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting transition kind: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetTransitionSettings tool
  server.tool(
    "obs-set-transition-settings",
    "Set the settings of the current transition",
    {
      transitionSettings: z.record(z.any()).describe("The settings to apply to the transition")
    },
    async ({ transitionSettings }) => {
      try {
        await client.sendRequest("SetCurrentSceneTransitionSettings", { transitionSettings });
        return {
          content: [
            {
              type: "text",
              text: "Successfully updated current transition settings"
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting transition settings: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetTransitionSettings tool
  server.tool(
    "obs-get-transition-settings",
    "Get the settings of the current transition",
    {},
    async () => {
      try {
        const settings = await client.sendRequest("GetCurrentSceneTransitionSettings");
        return {
          content: [
            {
              type: "text",
              text: JSON.stringify(settings, null, 2)
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting transition settings: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // TriggerStudioModeTransition tool
  server.tool(
    "obs-trigger-transition",
    "Trigger a scene transition in OBS (Studio Mode must be enabled)",
    {},
    async () => {
      try {
        await client.sendRequest("TriggerStudioModeTransition");
        return {
          content: [
            {
              type: "text",
              text: "Successfully triggered studio mode transition"
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error triggering transition: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );
}

File: py_src/transitions.py
#!/usr/bin/env python3

from typing import Any, Dict, List, Optional

from .client import obs_client
from .server import mcp

@mcp.tool()
async def get_transition_kind_list() -> List[str]:
    """
    Gets an array of all available transition kinds.
    
    Returns:
        List of transition kinds
    """
    response = await obs_client.send_request("GetTransitionKindList")
    return response.get("transitionKinds", [])

@mcp.tool()
async def get_scene_transition_list() -> Dict[str, Any]:
    """
    Gets an array of all scene transitions in OBS.
    
    Returns:
        Dict containing:
        - currentSceneTransitionKind: Kind of the current scene transition
        - currentSceneTransitionName: Name of the current scene transition
        - currentSceneTransitionDuration: Duration of the current scene transition (in milliseconds)
        - transitions: Array of transitions (each with name, kind)
    """
    return await obs_client.send_request("GetSceneTransitionList")

@mcp.tool()
async def get_current_scene_transition() -> Dict[str, Any]:
    """
    Gets information about the current scene transition.
    
    Returns:
        Dict containing transition information including kind, name, duration, and settings
    """
    return await obs_client.send_request("GetCurrentSceneTransition")

@mcp.tool()
async def set_current_scene_transition(transition_name: str) -> None:
    """
    Sets the current scene transition.
    
    Args:
        transition_name: Name of the transition to set as current
    """
    await obs_client.send_request("SetCurrentSceneTransition", {"transitionName": transition_name})

@mcp.tool()
async def set_current_scene_transition_duration(transition_duration: int) -> None:
    """
    Sets the duration of the current scene transition (if supported).
    
    Args:
        transition_duration: Duration in milliseconds
    """
    await obs_client.send_request("SetCurrentSceneTransitionDuration", 
                                 {"transitionDuration": transition_duration})

@mcp.tool()
async def set_current_scene_transition_settings(transition_settings: Dict[str, Any], overlay: bool = True) -> None:
    """
    Sets the settings of the current scene transition.
    
    Args:
        transition_settings: Settings object to apply to the transition
        overlay: Whether to overlay with existing settings or replace them
    """
    await obs_client.send_request("SetCurrentSceneTransitionSettings", {
        "transitionSettings": transition_settings,
        "overlay": overlay
    })

@mcp.tool()
async def get_scene_transition_override(scene_name: str) -> Dict[str, Any]:
    """
    Gets the scene transition override for a scene.
    
    Args:
        scene_name: Name of the scene
    
    Returns:
        Dict containing transition name and duration (if override exists)
    """
    return await obs_client.send_request("GetSceneSceneTransitionOverride", {"sceneName": scene_name})

@mcp.tool()
async def set_scene_transition_override(scene_name: str, transition_name: Optional[str] = None, 
                                       transition_duration: Optional[int] = None) -> None:
    """
    Sets the scene transition override for a scene.
    
    Args:
        scene_name: Name of the scene
        transition_name: Name of the transition to use, or null to remove
        transition_duration: Duration in milliseconds of the transition, or null to use default
    """
    payload = {"sceneName": scene_name}
    
    if transition_name is not None:
        payload["transitionName"] = transition_name
    
    if transition_duration is not None:
        payload["transitionDuration"] = transition_duration
    
    await obs_client.send_request("SetSceneSceneTransitionOverride", payload)

@mcp.tool()
async def trigger_studio_mode_transition() -> None:
    """
    Triggers the current scene transition. Only available when studio mode is enabled.
    """
    await obs_client.send_request("TriggerStudioModeTransition")

@mcp.tool()
async def set_tbar_position(tbar_position: float) -> None:
    """
    Sets the position of the transition bar.
    
    Args:
        tbar_position: Position to set the T-bar to (0.0-1.0)
    """
    await obs_client.send_request("SetTBarPosition", {"position": tbar_position})